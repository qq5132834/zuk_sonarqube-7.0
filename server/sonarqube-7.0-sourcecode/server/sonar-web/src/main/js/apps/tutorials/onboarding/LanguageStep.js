/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import NewProjectForm from './NewProjectForm';
import RadioToggle from '../../../components/controls/RadioToggle';
import { translate } from '../../../helpers/l10n';

/*::
type Props = {|
  onDone: (result: Result) => void,
  onReset: () => void,
  organization?: string,
  sonarCloud: boolean
|};
*/

/*::
type State = {
  language?: string,
  javaBuild?: string,
  cFamilyCompiler?: string,
  os?: string,
  projectKey?: string
};
*/

/*::
export type Result = State; */

export default class LanguageStep extends React.PureComponent {
  /*:: props: Props; */

  static defaultProps = { sonarCloud: false };

  state /*: State */ = {};

  isConfigured = () => {
    const { language, javaBuild, cFamilyCompiler, os, projectKey } = this.state;
    const isJavaConfigured = language === 'java' && javaBuild != null;
    const isDotNetConfigured = language === 'dotnet' && projectKey != null;
    const isCFamilyConfigured =
      language === 'c-family' && (cFamilyCompiler === 'msvc' || os != null) && projectKey != null;
    const isOtherConfigured = language === 'other' && projectKey != null;

    return isJavaConfigured || isDotNetConfigured || isCFamilyConfigured || isOtherConfigured;
  };

  handleChange = () => {
    if (this.isConfigured()) {
      this.props.onDone(this.state);
    } else {
      this.props.onReset();
    }
  };

  handleLanguageChange = (language /*: string */) => {
    this.setState({ language }, this.handleChange);
  };

  handleJavaBuildChange = (javaBuild /*: string */) => {
    this.setState({ javaBuild }, this.handleChange);
  };

  handleCFamilyCompilerChange = (cFamilyCompiler /*: string */) => {
    this.setState({ cFamilyCompiler }, this.handleChange);
  };

  handleOSChange = (os /*: string */) => {
    this.setState({ os }, this.handleChange);
  };

  handleProjectKeyDone = (projectKey /*: string */) => {
    this.setState({ projectKey }, this.handleChange);
  };

  handleProjectKeyDelete = () => {
    this.setState({ projectKey: undefined }, this.handleChange);
  };

  renderJavaBuild = () => (
    <div className="big-spacer-top">
      <h4 className="spacer-bottom">{translate('onboarding.language.java.build_technology')}</h4>
      <RadioToggle
        name="java-build"
        onCheck={this.handleJavaBuildChange}
        options={['maven', 'gradle'].map(build => ({
          label: translate('onboarding.language.java.build_technology', build),
          value: build
        }))}
        value={this.state.javaBuild}
      />
    </div>
  );

  renderCFamilyCompiler = () => (
    <div className="big-spacer-top">
      <h4 className="spacer-bottom">{translate('onboarding.language.c-family.compiler')}</h4>
      <RadioToggle
        name="c-family-compiler"
        onCheck={this.handleCFamilyCompilerChange}
        options={['msvc', 'clang-gcc'].map(compiler => ({
          label: translate('onboarding.language.c-family.compiler', compiler),
          value: compiler
        }))}
        value={this.state.cFamilyCompiler}
      />
    </div>
  );

  renderOS = () => (
    <div className="big-spacer-top">
      <h4 className="spacer-bottom">{translate('onboarding.language.os')}</h4>
      <RadioToggle
        name="os"
        onCheck={this.handleOSChange}
        options={['linux', 'win', 'mac'].map(os => ({
          label: translate('onboarding.language.os', os),
          value: os
        }))}
        value={this.state.os}
      />
    </div>
  );

  renderProjectKey = () => (
    <NewProjectForm
      onDelete={this.handleProjectKeyDelete}
      onDone={this.handleProjectKeyDone}
      organization={this.props.organization}
      projectKey={this.state.projectKey}
    />
  );

  render() {
    const shouldAskProjectKey =
      this.state.language === 'dotnet' ||
      (this.state.language === 'c-family' &&
        (this.state.cFamilyCompiler === 'msvc' ||
          (this.state.cFamilyCompiler === 'clang-gcc' && this.state.os != null))) ||
      (this.state.language === 'other' && this.state.os !== undefined);

    const languages = this.props.sonarCloud
      ? ['java', 'dotnet', 'c-family', 'other']
      : ['java', 'dotnet', 'other'];

    return (
      <div>
        <div>
          <h4 className="spacer-bottom">{translate('onboarding.language')}</h4>
          <RadioToggle
            name="language"
            onCheck={this.handleLanguageChange}
            options={languages.map(language => ({
              label: translate('onboarding.language', language),
              value: language
            }))}
            value={this.state.language}
          />
        </div>
        {this.state.language === 'java' && this.renderJavaBuild()}
        {this.state.language === 'c-family' && this.renderCFamilyCompiler()}
        {((this.state.language === 'c-family' && this.state.cFamilyCompiler === 'clang-gcc') ||
          this.state.language === 'other') &&
          this.renderOS()}
        {shouldAskProjectKey && this.renderProjectKey()}
      </div>
    );
  }
}
