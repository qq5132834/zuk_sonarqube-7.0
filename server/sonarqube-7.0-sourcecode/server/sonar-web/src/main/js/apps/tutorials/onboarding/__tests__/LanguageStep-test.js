/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import { shallow } from 'enzyme';
import LanguageStep from '../LanguageStep';

it('selects java', () => {
  const onDone = jest.fn();
  const wrapper = shallow(<LanguageStep onDone={onDone} onReset={jest.fn()} />);

  wrapper.find('RadioToggle').prop('onCheck')('java');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper
    .find('RadioToggle')
    .at(1)
    .prop('onCheck')('maven');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
  expect(onDone).lastCalledWith({ language: 'java', javaBuild: 'maven' });

  wrapper
    .find('RadioToggle')
    .at(1)
    .prop('onCheck')('gradle');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
  expect(onDone).lastCalledWith({ language: 'java', javaBuild: 'gradle' });
});

it('selects c#', () => {
  const onDone = jest.fn();
  const wrapper = shallow(<LanguageStep onDone={onDone} onReset={jest.fn()} />);

  wrapper.find('RadioToggle').prop('onCheck')('dotnet');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper.find('NewProjectForm').prop('onDone')('project-foo');
  expect(onDone).lastCalledWith({ language: 'dotnet', projectKey: 'project-foo' });
});

it('selects c-family', () => {
  const onDone = jest.fn();
  const wrapper = shallow(<LanguageStep onDone={onDone} onReset={jest.fn()} sonarCloud={true} />);

  wrapper.find('RadioToggle').prop('onCheck')('c-family');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper
    .find('RadioToggle')
    .at(1)
    .prop('onCheck')('msvc');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper.find('NewProjectForm').prop('onDone')('project-foo');
  expect(onDone).lastCalledWith({
    language: 'c-family',
    cFamilyCompiler: 'msvc',
    projectKey: 'project-foo'
  });

  wrapper
    .find('RadioToggle')
    .at(1)
    .prop('onCheck')('clang-gcc');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper
    .find('RadioToggle')
    .at(2)
    .prop('onCheck')('linux');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper.find('NewProjectForm').prop('onDone')('project-foo');
  expect(onDone).lastCalledWith({
    language: 'c-family',
    cFamilyCompiler: 'clang-gcc',
    os: 'linux',
    projectKey: 'project-foo'
  });
});

it('selects other', () => {
  const onDone = jest.fn();
  const wrapper = shallow(<LanguageStep onDone={onDone} onReset={jest.fn()} />);

  wrapper.find('RadioToggle').prop('onCheck')('other');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper
    .find('RadioToggle')
    .at(1)
    .prop('onCheck')('mac');
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  wrapper.find('NewProjectForm').prop('onDone')('project-foo');
  expect(onDone).lastCalledWith({ language: 'other', os: 'mac', projectKey: 'project-foo' });
});
