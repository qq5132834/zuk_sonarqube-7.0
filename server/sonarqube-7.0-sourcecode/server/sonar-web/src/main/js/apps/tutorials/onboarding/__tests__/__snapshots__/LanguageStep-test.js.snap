// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selects c# 1`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="dotnet"
    />
  </div>
  <NewProjectForm
    onDelete={[Function]}
    onDone={[Function]}
  />
</div>
`;

exports[`selects c-family 1`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.c-family",
            "value": "c-family",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="c-family"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.c-family.compiler
    </h4>
    <RadioToggle
      disabled={false}
      name="c-family-compiler"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.c-family.compiler.msvc",
            "value": "msvc",
          },
          Object {
            "label": "onboarding.language.c-family.compiler.clang-gcc",
            "value": "clang-gcc",
          },
        ]
      }
      value={null}
    />
  </div>
</div>
`;

exports[`selects c-family 2`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.c-family",
            "value": "c-family",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="c-family"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.c-family.compiler
    </h4>
    <RadioToggle
      disabled={false}
      name="c-family-compiler"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.c-family.compiler.msvc",
            "value": "msvc",
          },
          Object {
            "label": "onboarding.language.c-family.compiler.clang-gcc",
            "value": "clang-gcc",
          },
        ]
      }
      value="msvc"
    />
  </div>
  <NewProjectForm
    onDelete={[Function]}
    onDone={[Function]}
  />
</div>
`;

exports[`selects c-family 3`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.c-family",
            "value": "c-family",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="c-family"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.c-family.compiler
    </h4>
    <RadioToggle
      disabled={false}
      name="c-family-compiler"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.c-family.compiler.msvc",
            "value": "msvc",
          },
          Object {
            "label": "onboarding.language.c-family.compiler.clang-gcc",
            "value": "clang-gcc",
          },
        ]
      }
      value="clang-gcc"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.os
    </h4>
    <RadioToggle
      disabled={false}
      name="os"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.os.linux",
            "value": "linux",
          },
          Object {
            "label": "onboarding.language.os.win",
            "value": "win",
          },
          Object {
            "label": "onboarding.language.os.mac",
            "value": "mac",
          },
        ]
      }
      value={null}
    />
  </div>
</div>
`;

exports[`selects c-family 4`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.c-family",
            "value": "c-family",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="c-family"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.c-family.compiler
    </h4>
    <RadioToggle
      disabled={false}
      name="c-family-compiler"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.c-family.compiler.msvc",
            "value": "msvc",
          },
          Object {
            "label": "onboarding.language.c-family.compiler.clang-gcc",
            "value": "clang-gcc",
          },
        ]
      }
      value="clang-gcc"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.os
    </h4>
    <RadioToggle
      disabled={false}
      name="os"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.os.linux",
            "value": "linux",
          },
          Object {
            "label": "onboarding.language.os.win",
            "value": "win",
          },
          Object {
            "label": "onboarding.language.os.mac",
            "value": "mac",
          },
        ]
      }
      value="linux"
    />
  </div>
  <NewProjectForm
    onDelete={[Function]}
    onDone={[Function]}
    projectKey="project-foo"
  />
</div>
`;

exports[`selects java 1`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="java"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.java.build_technology
    </h4>
    <RadioToggle
      disabled={false}
      name="java-build"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java.build_technology.maven",
            "value": "maven",
          },
          Object {
            "label": "onboarding.language.java.build_technology.gradle",
            "value": "gradle",
          },
        ]
      }
      value={null}
    />
  </div>
</div>
`;

exports[`selects java 2`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="java"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.java.build_technology
    </h4>
    <RadioToggle
      disabled={false}
      name="java-build"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java.build_technology.maven",
            "value": "maven",
          },
          Object {
            "label": "onboarding.language.java.build_technology.gradle",
            "value": "gradle",
          },
        ]
      }
      value="maven"
    />
  </div>
</div>
`;

exports[`selects java 3`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="java"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.java.build_technology
    </h4>
    <RadioToggle
      disabled={false}
      name="java-build"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java.build_technology.maven",
            "value": "maven",
          },
          Object {
            "label": "onboarding.language.java.build_technology.gradle",
            "value": "gradle",
          },
        ]
      }
      value="gradle"
    />
  </div>
</div>
`;

exports[`selects other 1`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="other"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.os
    </h4>
    <RadioToggle
      disabled={false}
      name="os"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.os.linux",
            "value": "linux",
          },
          Object {
            "label": "onboarding.language.os.win",
            "value": "win",
          },
          Object {
            "label": "onboarding.language.os.mac",
            "value": "mac",
          },
        ]
      }
      value={null}
    />
  </div>
</div>
`;

exports[`selects other 2`] = `
<div>
  <div>
    <h4
      className="spacer-bottom"
    >
      onboarding.language
    </h4>
    <RadioToggle
      disabled={false}
      name="language"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.java",
            "value": "java",
          },
          Object {
            "label": "onboarding.language.dotnet",
            "value": "dotnet",
          },
          Object {
            "label": "onboarding.language.other",
            "value": "other",
          },
        ]
      }
      value="other"
    />
  </div>
  <div
    className="big-spacer-top"
  >
    <h4
      className="spacer-bottom"
    >
      onboarding.language.os
    </h4>
    <RadioToggle
      disabled={false}
      name="os"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "label": "onboarding.language.os.linux",
            "value": "linux",
          },
          Object {
            "label": "onboarding.language.os.win",
            "value": "win",
          },
          Object {
            "label": "onboarding.language.os.mac",
            "value": "mac",
          },
        ]
      }
      value="mac"
    />
  </div>
  <NewProjectForm
    onDelete={[Function]}
    onDone={[Function]}
  />
</div>
`;
